File: oas/CISCO/WordFinder.java
package oas.CISCO;

import java.util.ArrayList;
import java.util.List;

public class WordFinder {

    public static void main(String[] args) {

        char[][] grid = { { 'C', 'A', 'T' }, { 'I', 'D', 'O' }, { 'M', 'O', 'M' } };
        String[] words = "TAC TOM ADO MOM CLM KLM".split(" ");
        String[] result = getWords(grid, words);
        for (String res : result) {
            System.out.println(res);
        }
    }

    public static String[] getWords(char[][] letters, String[] words) {

        int m = letters.length;
        int n = letters[0].length;
        List<String> list = new ArrayList<>();

        for (int i = 0; i < m; i++) {
            StringBuffer buffer = new StringBuffer();
            for (int j = 0; j < n; j++) {
                buffer.append(letters[i][j]);
            }
            list.add(buffer.toString());
        }

        for (int i = 0; i < n; i++) {
            StringBuffer buffer = new StringBuffer();
            for (int j = 0; j < m; j++) {
                buffer.append(letters[j][i]);
            }
            list.add(buffer.toString());
        }

        String[] result = new String[words.length];
        int k = 0;
        for (int i = 0; i < words.length; i++) {
            boolean found = false;
            for (int j = 0; j < list.size(); j++) {
                StringBuffer buffer = new StringBuffer(list.get(j));
                if (buffer.toString().contains(words[i]) || buffer.reverse().toString().contains(words[i])) {
                    result[k++] = "Yes";
                    found = true;
                    break;
                }
            }
            if (!found) {
                result[k++] = "No";
            }
        }
        return result;
    }
}

File: python/oas/cisco/word_finder.py
from typing import List


class word_finder:
    def word_find(self, letters: List[List], words: List) -> List:
        m = len(letters)
        n = len(letters[0])
        
        lst = []
        
        for i in range(m):
            buffer = ""
            for j in range(n):
                buffer = buffer + letters[i][j]
            lst.append(buffer)
            
        for i in range(n):
            buffer = ""
            for j in range(m):
                buffer = buffer + letters[i][j]
            lst.append(buffer)

        result = ["None"] * len(words)
        k = 0
        for i in range(len(words)):
            found = False
            for j in range(len(lst)):
                buffer = lst[j]
                if words[i] in buffer or words[i] in reversed(buffer):
                    result[k] = "Yes"
                    k += 1
                    found = True
                    break
            if not found:
                result[k] = "No"
                k += 1
        return result

grid = [ [ 'C', 'A', 'T' ], [ 'I', 'D', 'O' ], [ 'M', 'O', 'M' ] ]
words = "CAT TOM ADO MOM CDM".split(" ")
print(word_finder().word_find(grid, words))     